### PR TITLE
Add Self-Description endpoint

### PR DESCRIPTION
**Pull request by bscholtes1A**
## What this PR changes/adds

Adds an endpoint serving Self-Description.

## Why it does that

See https://github.com/agera-edc/IdentityHub/issues/38. 

## Linked Issue(s)

https://github.com/agera-edc/IdentityHub/issues/38
https://github.com/agera-edc/IdentityHub/issues/11

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/identityservice/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/identityservice/blob/main/styleguide.md) for details_)

### CODE REVIEW

**Comment by bscholtes1A** [comment]
All tests are OK locally. Don't know why CICD is failing.

File: extensions/identity-hub/src/main/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtension.java [hunk 1]
```diff
@@ -49,12 +61,28 @@ public void initialize(ServiceExtensionContext context) {
...
+
+    /**
+     * Load content from a static JSON file.
+     *
+     * @param path Path to the static file.
+     * @return JSON file content.
+     */
+    private static JsonNode loadJsonFile(ObjectMapper mapper, String path) {
+        try {
+            var content = Files.readString(Path.of(path));
```
Comment (marcgs) [inline extensions/identity-hub/src/main/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtension.java]: As the self description is just a static file at this stage, why not just adding the file to the extension JAR and load it here from the classpath?

**Comment by marcgs** [comment]
> All tests are OK locally. Don't know why CICD is failing.

The test failing in CI is the system test:
```
VerifiableCredentialsIntegrationTest > push_and_get_verifiable_credentials(CredentialsVerifier, DidResolverRegistry) STANDARD_OUT
    INFO 2022-08-01T12:21:34.955721433 Initialized Jetty Service
    INFO 2022-08-01T12:21:34.958441467 Initialized Jersey Web Service
    INFO 2022-08-01T12:21:34.962473418 Initialized Identity Did Core
    INFO 2022-08-01T12:21:34.963818735 Initialized Identity Hub DID extension
    INFO 2022-08-01T12:21:34.964367442 Initialized org.eclipse.dataspaceconnector.identityhub.verifier.CredentialsVerifierExtension
    SEVERE 2022-08-01T12:21:34.968668996 Error booting runtime
    org.eclipse.dataspaceconnector.spi.EdcException: No setting found for key edc.self.description.path
```

This one does need a special setup to run locally (see [README](https://github.com/agera-edc/IdentityHub/blob/main/system-tests/README.md)), otherwise it is skipped.

The issue is that the system test is not provided with any self description document. My suggestion is (as mentioned in my previous comment) to add the self-description to the classpath and load it from there.

File: extensions/identity-hub/src/main/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtension.java (hunk 1 above)
Comment (bscholtes1A) [reply extensions/identity-hub/src/main/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtension.java]: As discussed, we keep the EDC setting enabling to provide a custom self-description. 
If this setting is not specified, then a static self-description from the classpath is used.

**Comment by bscholtes1A** [comment]
> > All tests are OK locally. Don't know why CICD is failing.
> 
> The test failing in CI is the system test:
> 
> ```
> VerifiableCredentialsIntegrationTest > push_and_get_verifiable_credentials(CredentialsVerifier, DidResolverRegistry) STANDARD_OUT
>     INFO 2022-08-01T12:21:34.955721433 Initialized Jetty Service
>     INFO 2022-08-01T12:21:34.958441467 Initialized Jersey Web Service
>     INFO 2022-08-01T12:21:34.962473418 Initialized Identity Did Core
>     INFO 2022-08-01T12:21:34.963818735 Initialized Identity Hub DID extension
>     INFO 2022-08-01T12:21:34.964367442 Initialized org.eclipse.dataspaceconnector.identityhub.verifier.CredentialsVerifierExtension
>     SEVERE 2022-08-01T12:21:34.968668996 Error booting runtime
>     org.eclipse.dataspaceconnector.spi.EdcException: No setting found for key edc.self.description.path
> ```
> 
> This one does need a special setup to run locally (see [README](https://github.com/agera-edc/IdentityHub/blob/main/system-tests/README.md)), otherwise it is skipped.
> 
> The issue is that the system test is not provided with any self description document. My suggestion is (as mentioned in my previous comment) to add the self-description to the classpath and load it from there.

Thanks! Should be fixed now.

File: extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtension.java [hunk 2]
```diff
@@ -21,16 +21,15 @@
 import org.eclipse.dataspaceconnector.identityhub.credentials.VerifiableCredentialsJwtServiceImpl;
```
Comment (bscholtes1A) [inline extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtension.java:21]: @marcgs @algattik 
Finally managed to make everything compile by refactoring the current extension in a more standard way. Creating @Provider methods depending on injected fields should be avoided in my opinion. 

@paullatzelsperger WDYT?

File: extensions/identity-hub/src/main/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtension.java [hunk 3]
```diff
@@ -49,12 +63,38 @@ public void initialize(ServiceExtensionContext context) {
         methodProcessorFactory.register(COLLECTIONS_WRITE, new CollectionsWriteProcessor(identityHubStore));
         methodProcessorFactory.register(FEATURE_DETECTION_READ, new FeatureDetectionReadProcessor());
 
-        var identityHubController = new IdentityHubController(methodProcessorFactory);
+        var path = context.getSetting(SELF_DESCRIPTION_DOCUMENT_PATH_SETTING, null);
+        JsonNode selfDescription;
+        try (var is = loadSelfDescriptionDocument(path)) {
```
Comment (marcgs) [inline extensions/identity-hub/src/main/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtension.java]: I would move the whole logic to load the self description as a JsonNode into the method, or even better, into a collaborator `SelfDescriptionLoader` that can be unit tested?

File: extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtensionTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(DependencyInjectionExtension.class)
```
Comment (marcgs) [inline extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtensionTest.java:40]: We have already a `IdentityHubControllerTest`, why did not add a test for the self-description there?

File: extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtensionTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+class IdentityHubExtensionTest {
```
Comment (marcgs) [inline extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/IdentityHubExtensionTest.java]: This test is a bit smelly. It is actually not testing the extension, but rather the initialisation of the self-description document but in a convoluted way. I think if we extract the logic for loading the self-description document into a separate class, we could unit test that class nicely and without mixing it with other responsibilities.

File: .github/actions/gradle-setup/action.yml [hunk 6]
```diff
@@ -3,41 +3,41 @@ description: "Setup Gradle"
 runs:
```
Comment (bscholtes1A) [inline .github/actions/gradle-setup/action.yml]: Temporary change for testing with CICD. 
This file will be restored when https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/1792 will be merged

File: .github/actions/gradle-setup/action.yml [hunk 7]
```diff
@@ -3,41 +3,41 @@ description: "Setup Gradle"
...
   using: "composite"
   steps:
-    # Checkout EDC code into DataSpaceConnector directory.
```
Comment (algattik) [inline .github/actions/gradle-setup/action.yml]: why this change? Though both forms are valid, it is quite more common in my experience to indent there (also in GH doc)

Comment (bscholtes1A) [reply .github/actions/gradle-setup/action.yml]: As mentioned above, this file will be restored as soon as https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/1792#pullrequestreview-1065206083 gets merged

File: .github/actions/gradle-setup/action.yml (hunk 6 above)
Comment (algattik) [reply .github/actions/gradle-setup/action.yml]: Are you planning to (perhaps cherry pick and)publish an EDC snapshot?

File: client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java [hunk 8]
```diff
@@ -151,11 +171,15 @@ void add_invalidPrivateKey_fails() throws JsonProcessingException {
...
 
         // assert
-        assertThat(exitCode).isNotEqualTo(0);
+        assertThat(exitCode).isNotZero();
         assertThat(outContent).isEmpty();
         assertThat(errContent).contains("Error while signing Verifiable Credential");
     }
 
+    private int executeGetSelfDescription() {
+        return cmd.execute("-s", HUB_URL, "sd");
```
Comment (marcgs) [inline client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java]: Consider adding a suboption for the command like "sd get". I assume we will add "sd upload" in the future.

**Review by marcgs — APPROVED** [review]

File: client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java (hunk 8 above)
Comment (bscholtes1A) [reply client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java]: Done in https://github.com/agera-edc/IdentityHub/pull/37/commits/14d238e422c0e272ec643caa5aaa69820a6f4955

**Comment by paullatzelsperger** [comment]
@bscholtes1A I'll remove myself, since there are already two reviewers on it. If you still want me to take a look, please just re-request.

File: client-cli/src/main/java/org/eclipse/dataspaceconnector/identityhub/cli/GetSelfDescriptionCommand.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    private CommandLine.Model.CommandSpec spec;
+
+    @Override
+    public Integer call() throws Exception {
+        var out = spec.commandLine().getOut();
+        var result = command.cli.identityHubClient.getSelfDescription(command.cli.hubUrl);
+        if (result.succeeded()) {
+            MAPPER.writeValue(out, result.getContent());
+            out.println();
+        } else {
```
Comment (algattik) [inline client-cli/src/main/java/org/eclipse/dataspaceconnector/identityhub/cli/GetSelfDescriptionCommand.java]: code flow can be simplified by removing else (reduce number of code branches)

File: client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java [hunk 10]
```diff
@@ -71,6 +71,25 @@ void setUp() {
...
+        // act
+        var exitCode = executeGetSelfDescription();
+        var outContent = out.toString();
+        var errContent = err.toString();
+
+        // assert
+        assertThat(exitCode).isZero();
+        assertThat(errContent).isEmpty();
+
+        assertThat(outContent).isEqualToIgnoringWhitespace(selfDescription.toString());
```
Comment (algattik) [inline client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java]: why ignoring whitespace? If we expect the JSON to be seen by users, then things like multiline and final EOL are not ancillary, but part of the test scope.

File: extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java [hunk 11]
```diff
@@ -14,66 +14,68 @@
...
-        when(publicKeyResolver.resolvePublicKey(anyString())).thenReturn(Result.success(toPublicKeyWrapper(jwk)));
-        var didDocument = DidDocument.Builder.newInstance()
-                .id(SUBJECT)
-                .service(List.of(new Service("IdentityHub", "IdentityHub", API_URL)))
-                .build();
+        var didDocument = createDidDocument(jwk);
         var credential = generateVerifiableCredential();
+        var method = FAKER.lorem().word();
         var jwt = buildSignedJwt(credential, CREDENTIAL_ISSUER, SUBJECT, jwk);
+        registry.register(new MockDidResolver(DID_METHOD, didDocument));
```
Comment (algattik) [inline extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java]: why not use of mockito?

**Review by algattik — APPROVED** [review]

File: client-cli/src/main/java/org/eclipse/dataspaceconnector/identityhub/cli/GetSelfDescriptionCommand.java (hunk 9 above)
Comment (bscholtes1A) [reply client-cli/src/main/java/org/eclipse/dataspaceconnector/identityhub/cli/GetSelfDescriptionCommand.java]: Updated

File: extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java (hunk 11 above)
Comment (bscholtes1A) [reply extensions/identity-hub-verifier/src/test/java/org/eclipse/dataspaceconnector/identityhub/verifier/CredentialsVerifierExtensionTest.java]: Done

File: client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java (hunk 10 above)
Comment (bscholtes1A) [reply client-cli/src/test/java/org/eclipse/dataspaceconnector/identityhub/cli/VerifiableCredentialsCommandTest.java]: Updated
